mcos/de_noise.py

```diff
@@ -0,0 +1,179 @@
+    sse = np.sum((empirical_pdf - theoretical_pdf) ** 2)
+    return sse
+
+
+def _mp_PDF(var: float, q: float, pts: int) -> pd.Series:
+    """
+    Creates a theoretical Marcenko-Pastur probability density function
+    :param var: variance 𝜎^2
+    :param q: q=T/N where T=sample length and N=number of variables
+    :param pts: number of points in the distribution
+    :return: a theoretical Marcenko-Pastur probability density function
+    """
+    min_eigenvalue, max_eigenvalue = var * (1 - (1. / q) ** .5) ** 2, var * (1 + (1. / q) ** .5) ** 2
+    eigenvalues = np.linspace(min_eigenvalue, max_eigenvalue, pts).flatten()
+    pdf = q / (2 * np.pi * var * eigenvalues) * ((max_eigenvalue - eigenvalues) * (eigenvalues - min_eigenvalue)) ** .5
+    pdf = pdf.flatten()
+    pdf = pd.Series(pdf, index=eigenvalues)
+    return pdf
+
+
+def _fit_KDE(obs: np.array, bandwidth: float = .25, kernel: str = 'gaussian', x: np.array = None) -> pd.Series:
+    """
+    Fit kernel to a series of observations, and derive the prob of observations.
+    x is the array of values on which the fit KDE will be evaluated
+    :param obs: the series of observations
+    :param bandwidth: bandwidth hyper-parameter for KernelDensity
+    :param kernel: kernel hyper-parameter for KernelDensity
+    :param x: array of values _fit_KDE will be evaluated against
+    :return: an empirical Marcenko-Pastur probability density function
+    """
+    if len(obs.shape) == 1:
+        obs = obs.reshape(-1, 1)
+    kde = KernelDensity(kernel=kernel, bandwidth=bandwidth).fit(obs)
+    if x is None:
+        x = np.unique(obs).reshape(-1, 1)
+    if len(x.shape) == 1:
+        x = x.reshape(-1, 1)
+    log_prob = kde.score_samples(x)  # log(density)
+    pdf = pd.Series(np.exp(log_prob), index=x.flatten())
+    return pdf
+
+
+def _de_noised_corr(eigenvalues: np.array, eigenvectors: np.array, n_facts: int) -> np.array:
+    """
+    Shrinks the eigenvalues associated with noise, and returns a de-noised correlation matrix
+    :param eigenvalues: array of eigenvalues
+    :param eigenvectors: array of eigenvectors
+    :param n_facts: number of elements in diagonalized eigenvalues to replace with the mean of eigenvalues
+    :return: de-noised correlation matrix
+    """
+    # Remove noise from corr by fixing random eigenvalues
+    eigenvalues_ = np.diag(eigenvalues).copy()
+    eigenvalues_[n_facts:] = eigenvalues_[n_facts:].sum() / float(eigenvalues_.shape[0] - n_facts)
+    eigenvalues_ = np.diag(eigenvalues_)
+    corr = np.dot(eigenvectors, eigenvalues_).dot(eigenvectors.T)
+    corr = _cov_to_corr(corr)
+    return corr
+
+
+def _corr_to_cov(corr: np.array, std: np.array) -> np.array:
+    """
+    Recovers the covariance matrix from the de-noise correlation matrix
+    :param corr: de-noised correlation matrix
+    :param std: standard deviation of the correlation matrix
+    :return: a recovered covariance matrix
+    """
+    cov = corr * np.outer(std, std)
+    return cov
```

Comment:
Are these functions all specific to `CovarianceMatrixDeNoiser`? If so could we move them inside the class?

Reply:
for sure, I was thinking that it'd be nice to have `de_noise_covariance_matrix` separate so we could use it elsewhere if needed, but it seems like `CovarianceMatrixDeNoiser` handles that anyways